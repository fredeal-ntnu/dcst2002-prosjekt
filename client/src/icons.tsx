import * as React from 'react';
import { Component } from 'react-simplified';

class BaseIcon extends Component<{ size?: number; color?: string; style?: React.CSSProperties }> {
  size = this.props.size || 16;
  color = this.props.color || 'currentColor';
  style = this.props.style || {};

  getSvg() {
    return '';
  }

  render() {
    const svgStyle = {
      verticalAlign: '-0.125em',
      ...this.style,
    };

    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width={this.size}
        height={this.size}
        fill={this.color}
        viewBox="0 0 16 16"
        style={svgStyle}
        dangerouslySetInnerHTML={{ __html: this.getSvg() }}
      />
    );
  }
}

// EyeIcon class
export class EyeIcon extends BaseIcon {
  getSvg() {
    return `
      <path d="M16 8s-3-5.5-8-5.5S0 8 0 8s3 5.5 8 5.5S16 8 16 8zM1.173 8a13.133 13.133 0 0 1 1.66-2.043C4.12 4.668 5.88 3.5 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.133 13.133 0 0 1 14.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755C11.879 11.332 10.119 12.5 8 12.5c-2.12 0-3.879-1.168-5.168-2.457A13.134 13.134 0 0 1 1.172 8z"/>
      <path d="M8 5.5a2.5 2.5 0 1 0 0 5 2.5 2.5 0 0 0 0-5zM4.5 8a3.5 3.5 0 1 1 7 0 3.5 3.5 0 0 1-7 0z"/>
    `;
  }
}

// SearchIcon class
export class SearchIcon extends BaseIcon {
  getSvg() {
    return `
      <path d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"/>
      `;
  }
}

// QuestionIcon class
export class QuestionIcon extends BaseIcon {
  getSvg() {
    return `
    <path d="M5 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm4 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm3 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2z"/>
    <path d="m2.165 15.803.02-.004c1.83-.363 2.948-.842 3.468-1.105A9.06 9.06 0 0 0 8 15c4.418 0 8-3.134 8-7s-3.582-7-8-7-8 3.134-8 7c0 1.76.743 3.37 1.97 4.6a10.437 10.437 0 0 1-.524 2.318l-.003.011a10.722 10.722 0 0 1-.244.637c-.079.186.074.394.273.362a21.673 21.673 0 0 0 .693-.125zm.8-3.108a1 1 0 0 0-.287-.801C1.618 10.83 1 9.468 1 8c0-3.192 3.004-6 7-6s7 2.808 7 6c0 3.193-3.004 6-7 6a8.06 8.06 0 0 1-2.088-.272 1 1 0 0 0-.711.074c-.387.196-1.24.57-2.634.893a10.97 10.97 0 0 0 .398-2z"/>
    `;
  }
}

// FavouriteIcon class
export class FavouriteIcon extends BaseIcon {
  getSvg() {
    return `
    <path d="m8 2.748-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"/>
    `;
  }
}

// UpvoteIcon class
export class UpvoteIcon extends BaseIcon {
  getSvg() {
    return `
<path d="M2 16a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v12a2 2 0 0 1-2 2H2zm6.5-4.5V5.707l2.146 2.147a.5.5 0 0 0 .708-.708l-3-3a.5.5 0 0 0-.708 0l-3 3a.5.5 0 1 0 .708.708L7.5 5.707V11.5a.5.5 0 0 0 1 0z"/>`;
  }
}

// DownvoteIcon class
export class DownvoteIcon extends BaseIcon {
  getSvg() {
    return `
<path d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm6.5 4.5v5.793l2.146-2.147a.5.5 0 0 1 .708.708l-3 3a.5.5 0 0 1-.708 0l-3-3a.5.5 0 1 1 .708-.708L7.5 10.293V4.5a.5.5 0 0 1 1 0z"/>`;
  }
}

// CommentBubleIcon class
export class CommentBubleIcon extends BaseIcon {
  getSvg() {
    return `
    <path d="M16 8c0 3.866-3.582 7-8 7a9.06 9.06 0 0 1-2.347-.306c-.584.296-1.925.864-4.181 1.234-.2.032-.352-.176-.273-.362.354-.836.674-1.95.77-2.966C.744 11.37 0 9.76 0 8c0-3.866 3.582-7 8-7s8 3.134 8 7zM5 8a1 1 0 1 0-2 0 1 1 0 0 0 2 0zm4 0a1 1 0 1 0-2 0 1 1 0 0 0 2 0zm3 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2z"/>
    `;
  }
}

//@ts-ignore
<svg
  xmlns="http://www.w3.org/2000/svg"
  width="16"
  height="16"
  fill="currentColor"
  class="bi bi-chat-dots"
  viewBox="0 0 16 16"
>
  <path d="M5 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm4 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm3 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2z" />
  <path d="m2.165 15.803.02-.004c1.83-.363 2.948-.842 3.468-1.105A9.06 9.06 0 0 0 8 15c4.418 0 8-3.134 8-7s-3.582-7-8-7-8 3.134-8 7c0 1.76.743 3.37 1.97 4.6a10.437 10.437 0 0 1-.524 2.318l-.003.011a10.722 10.722 0 0 1-.244.637c-.079.186.074.394.273.362a21.673 21.673 0 0 0 .693-.125zm.8-3.108a1 1 0 0 0-.287-.801C1.618 10.83 1 9.468 1 8c0-3.192 3.004-6 7-6s7 2.808 7 6c0 3.193-3.004 6-7 6a8.06 8.06 0 0 1-2.088-.272 1 1 0 0 0-.711.074c-.387.196-1.24.57-2.634.893a10.97 10.97 0 0 0 .398-2z" />
</svg>;
