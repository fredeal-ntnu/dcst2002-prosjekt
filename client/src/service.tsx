import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Question = {
  question_id: number;
  title: string;
  text: string;
  view_count: number;
  has_answer: boolean;
  user_id: string;
};

export type Tag = {
  tag_id: number;
  name: string;
};

export type Tag_Question_Relation = {
  tag_id: number;
  question_id: number;
};

export type User = {
  user_name: string;
  password: string;
};

export type Answer = {
  answer_id: number;
  text: string;
  confirmed_answer: boolean;
  question_id: number;
  score?: number;
};

export type QuestionComment = {
  question_comment_id: number;
  text: string;
  question_id: number;
};

export type AnswerComment = {
  answer_comment_id: number;
  text: string;
  answer_id: number;
};

export type Vote = {
  user_id: number;
  answer_id: number;
  vote_type: boolean;
}

class Service {
  /**
   * Get all questions.
   */

  getAllQuestions() {
    return axios.get<Question[]>('/questions').then((response) => response.data);
  }

  /**
   * Get a question by id.
   */

  getQuestion(id: number) {
    return axios.get<Question>('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get top five questions.
   */

  getTopFiveQuestions() {
    return axios.get<Question[]>('/topfivequestions').then((response) => response.data);
  }

  /**
   * Get all unaswered questions.
   */
  getUnansweredQuestions() {
    return axios.get<Question[]>('/unansweredquestions').then((response) => response.data);
  }

  /**
   * Update a question.
   */

  updateQuestion(question: Question) {
    return axios.put('/questions', question).then((response) => response.data);
  }

  /**
   * Create a question.
   */

  createQuestion(title: string, text: string, user_id: number) {
    return axios.post('/questions', { title, text, user_id }).then((response) => response.data.id);
  }

  /**
   * Delete a question.
   */

  deleteQuestion(id: number) {
    console.log(typeof id);
    return axios.delete('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get tags for a question.
   */

  getTags(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  /**
   * Get all tags.
   */

  getAllTags() {
    return axios.get('/tags').then((response) => response.data);
  }

  createTagQuestionRelation(tag_id: number, question_id: number) {
    return axios.post('/questions/' + question_id, { tag_id, question_id });
  }

  /**
   * Get a tag by name.
   */

  getQuestionsByTagId(tag: number) {
    return axios.get('/tag/' + tag + '/questions').then((response) => response.data);
  }

  //get all tags for a question by question id
  getTagsForQuestion(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  getAllTagQuestionsRelations() {
    return axios.get('/question/:id').then((response) => response.data);
  }

  //get a user
  getUser(user: User) {
    return axios.get<User>('/users/' + user.user_name).then((response) => response.data);
  }

  getAnswerById(id: number) {
    return axios.get('/answers/' + id).then((response) => response.data);
  }

  getAnswersForQuestion(id: number) {
    return axios.get('/questions/' + id + '/answers').then((response) => response.data);
  }

  createAnswer(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/answers', { text, question_id })
      .then((response) => response.data);
  }

  // ALLE SERVICES FOR QUESTION COMMENTS

  getQuestionCommentById(id: number) {
    return axios
    .get('/comments/'+ id)
    .then((response) => response.data);
  }

  createQuestionComment(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/comments', { text, question_id })
      .then((response) => response.data);
  }

updateQuestionComment(questionComment: QuestionComment) {
  return axios
  .put('/comments/', questionComment)
  .then((response) => response.data);
}

  deleteQuestionComment(id: number) {
    return axios
      .delete('/comments/' + id)
      .then((response) => response.data);
  }

  getQuestionCommentsForQuestion(id: number) {
    return axios
    .get('/questions/' + id + '/comments')
    .then((response) => response.data);
  }

  // ALLE SERVICES FOR ANSWER COMMENTS
  

  getAnswerCommentsForAnswer(id: number) {
    return axios.get('/answers/' + id + '/comments')
    .then((response) => response.data);
  }

  createAnswerComment(text: string, answer_id: number) {
    return axios
      .post('/answers/' + answer_id + '/comments', { text, answer_id })
      .then((response) => response.data);
  }

  getAnswerCommentById(id: number) {
    console.log('andre')
    return axios
    .get('/answer/comments/' + id)
    .then((response) => response.data);
  }

  updateAnswerComment(answerComment: AnswerComment) {
    return axios
    .put('/comments/', answerComment)
    .then((response) => response.data);
  }

  deleteAnswerComment(id: number) {
    return axios
      .delete('/answer/comments/' + id)
      .then((response) => response.data);
  }

  


  
  updateAnswer(answer: Answer) {
    return axios.put('/answers', answer)
    .then((response) => response.data);
  }

deleteAnswer(id: number) {
  return axios
  .delete('/answers/' + id)
  .then((response) => response.data);
}



// alle services for votes

getVotesByAnswerId(id: number) {
  return axios
  .get('/answers/' + id + '/votes')
  .then((response) => response.data);

}

createVoteForAnswer(user_id: number, answer_id: string, vote_type: boolean) {
  return axios
  .post(answer_id + '/votes', { user_id, answer_id, vote_type })
  .then((response) => response.data);
}


}

const service = new Service();
export default service;
