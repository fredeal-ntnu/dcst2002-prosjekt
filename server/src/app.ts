import express from 'express';
import questionRouter from './router/question-router';
import answerRouter from './router/answer-router';
import tagRouter from './router/tag-router';
import questionRelationRouter from './router/question-relation_router';
// import loginRouter from './router/login-router';
// import loginRouter from './router/login-router';
import questionCommentRouter from './router/question-comment-router';
 import answerCommentRouter from './router/answer-comment-router';
import voteRouter from './router/vote-router';
// import session from 'express-session';
// import passport from 'passport';
// import cookieParser from 'cookie-parser';

// const cookieparser = require('cookie-parser');

/**
 * Express application.
 */
const app = express();

app.use(express.json());

// Since API is not compatible with v1, API version is increased to v2
app.use('/api/v2', questionRouter);
app.use('/api/v2', tagRouter);
app.use('/api/v2', questionRelationRouter);
// app.use('/api/v2', loginRouter);
app.use('/api/v2', answerRouter);
app.use('/api/v2', questionCommentRouter);
app.use('/api/v2', answerCommentRouter);
app.use('/api/v2', voteRouter);

//app.use(cookieParser());
app.use(express.urlencoded({ extended: true }));

//serving public file
app.use(express.static(__dirname));

const oneDay = 1000 * 60 * 60 * 24;

// app.use(
//   session({
//     secret: 'secretKeyNumberTwoFiveFour',
//     saveUninitialized: true,
//     cookie: {
//       maxAge: oneDay,
//       httpOnly: true,
//       sameSite: true,
//     },
//     resave: false,
//   }),
// );

// app.use(passport.authenticate('session'));

export default app;
