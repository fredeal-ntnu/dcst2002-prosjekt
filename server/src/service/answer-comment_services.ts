import pool from '../mysql-pool';
import type { RowDataPacket, ResultSetHeader } from 'mysql2';

export type Answer_Comment_Content = {
  answer_comment_id: number;
  text: string;
  answer_id: number;
  user_id: number;
};

class Service {
  //get answer comment by answer_id

  getAnswerCommentsByAnswerId(answer_id: number) {
    return new Promise<Answer_Comment_Content[]>((resolve, reject) => {
      pool.query(
        'SELECT * FROM Answer_comments WHERE answer_id = ?',
        [answer_id],
        (error, results: RowDataPacket[]) => {
          if (error) return reject(error);

          resolve(results as Answer_Comment_Content[]);
        },
      );
    });
  }

  //Get answer comment by answer_comment_id

  getAnswerCommentById(answer_comment_id: number) {
    return new Promise<Answer_Comment_Content>((resolve, reject) => {
      pool.query(
        'SELECT * FROM Answer_comments WHERE Answer_comment_id = ?',
        [answer_comment_id],
        (error, results: RowDataPacket[]) => {
          if (error) return reject(error);
          if (results.length === 0) return reject('No answer comment found');

          resolve(results[0] as Answer_Comment_Content);
        },
      );
    });
  }

  //Create answer comment

  createAnswerComment(text: string, answer_id: number, user_id: number) {
    return new Promise<number>((resolve, reject) => {
      pool.query(
        'INSERT INTO Answer_comments SET text=?, answer_id=?, user_id=?',
        [text, answer_id, user_id],
        (error, results: ResultSetHeader) => {
          if (error) return reject(error);

          resolve(results.insertId);
        },
      );
    });
  }

  //Delete answer comment with answer id.

  deleteAnswerComment(answer_comment_id: number) {
    return new Promise<void>((resolve, reject) => {
      pool.query(
        'DELETE FROM Answer_comments WHERE Answer_comment_id = ?',
        [answer_comment_id],
        (error, results: ResultSetHeader) => {
          if (error) return reject(error);
          if (results.affectedRows == 0) return reject('No row deleted');

          resolve();
        },
      );
    });
  }

  //Update answer Comment with given id.

  updateAnswerComment(answer_comment: Answer_Comment_Content) {
    return new Promise<void>((resolve, reject) => {
      pool.query(
        'UPDATE Answer_comments SET text=? WHERE Answer_comment_id=?',
        [answer_comment.text, answer_comment.answer_comment_id],
        (error, _results) => {
          if (error) return reject(error);

          resolve();
        },
      );
    });
  }
}

export const answerCommentService = new Service();
