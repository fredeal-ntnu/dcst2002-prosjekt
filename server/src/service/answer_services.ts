
import pool from '../mysql-pool';
import type { RowDataPacket, ResultSetHeader } from 'mysql2';

export type Answer_Content = {
  answer_id: number;
  text: string;
  confirmed_answer: boolean;
  question_id: number;
};

class Service {

    getAnswersByQuestionId(question_id: number) {
        return new Promise<Answer_Content[]>((resolve, reject) => {
            pool.query('SELECT * FROM Answers WHERE question_id = ?', [question_id], (error, results: RowDataPacket[]) => {
            if (error) return reject(error);

            resolve(results as Answer_Content[]);
            });
        });
    }

    setConfirmedAnswer(answer_id: number) {
        return new Promise<void>((resolve, reject) => {
            pool.query('UPDATE Answers SET confirmed_answer = 1 WHERE answer_id = ?', [answer_id], (error, results: ResultSetHeader) => {
            if (error) return reject(error);
            if (results.affectedRows == 0) return reject(new Error('No row updated'));

            resolve();
            });
        }); 
    }

   getAnswerById(answer_id: number) {
    return new Promise<Answer_Content>((resolve, reject) => {
        pool.query('SELECT * FROM Answers WHERE answer_id = ?', [answer_id], (error, results: RowDataPacket[]) => {
        if (error) return reject(error);
        if (results.length == 0) return reject(new Error('No answer found'));

        resolve(results[0] as Answer_Content);
        });
    });
   }

    createAnswer(text: string, question_id: number) {
        return new Promise<number>((resolve, reject) => {
            pool.query('INSERT INTO Answers SET text=?, question_id=?', [text, question_id], (error, results: ResultSetHeader) => {
            if (error) return reject(error);
    
            resolve(results.insertId);
            });
        });

    }

    /**
     * Delete answer with given id.
     */

    deleteAnswer(id: number) {
        return new Promise<void>((resolve, reject) => {
            pool.query('DELETE FROM Answers WHERE answer_id=?', [id], (error) => {
                console.log(error)
            if (error) return reject(error);
            resolve();
            });
        });
    }
    /**
     * Update answer with given id.
     */
    
    updateAnswer(answer: Answer_Content) {
        return new Promise<void>((resolve, reject) => {
            pool.query('UPDATE Answers SET text=?, confirmed_answer=? WHERE answer_id=?', 
            [answer.text, answer.confirmed_answer, answer.answer_id],
             (error, results: ResultSetHeader) => {
            if (error) return reject(error);
            if (results.affectedRows == 0) return reject(new Error('No row updated'));

            resolve();
            });
        }); 
    }
    
}
export const answerService = new Service(); 
