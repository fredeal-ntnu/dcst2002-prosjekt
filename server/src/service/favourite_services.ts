import pool from '../mysql-pool';
import type { RowDataPacket, ResultSetHeader } from 'mysql2';

export type Favourite_Content = {
question_id: number;
user_id: number;
};



class Service {

    //get all favourites for user
    getAllFavouritesForUser(user_id: number) {
        return new Promise<Favourite_Content[]>((resolve, reject) => {
            pool.query(
            'SELECT * FROM question_user_favourites WHERE user_id = ?',
            [user_id],
            (error, results: RowDataPacket[]) => {
                if (error) return reject(error);
                
                resolve(results as Favourite_Content[]);
            },
            );
        });
    }



    //get all favourites by question id
    getAllFavouritesByQuestionId(question_id: number) {
        return new Promise<Favourite_Content[]>((resolve, reject) => {
            pool.query(
            'SELECT * FROM question_user_favourites WHERE question_id = ?',
            [question_id],
            (error, results: RowDataPacket[]) => {
                if (error) return reject(error);
    
                resolve(results as Favourite_Content[]);
            },
            );
        });
    }

    //create favourite relation with questionid and userid
    createFavouriteRelation(answer_id: number, user_id: number) {
      return new Promise<void>((resolve, reject) => {
          pool.query(
              'SELECT * FROM answer_user_favourite WHERE answer_id = ? AND user_id = ?',
              [answer_id, user_id],
              (selectError, results) => {
                  if (selectError) {
                      return reject(selectError);
                  }
  
                if ((results as RowDataPacket[]).length === 0) {
                      // If no matching row exists, insert a new one
                      pool.query(
                          'INSERT INTO answer_user_favourite (answer_id, user_id) VALUES (?, ?)',
                          [answer_id, user_id],
                          (insertError) => {
                              if (insertError) {
                                  return reject(insertError);
                              }
                              resolve();
                          }
                      );
                  } else {
                      // If a matching row exists, delete it
                      pool.query(
                          'DELETE FROM answer_user_favourite WHERE answer_id = ? AND user_id = ?',
                          [answer_id, user_id],
                          (deleteError) => {
                              if (deleteError) {
                                  return reject(deleteError);
                              }
                              resolve();
                          }
                      );
                  }
              }
          );
      });
  }
  


    //get all favourite relations
    getAllFavouriteRelations() {
        return new Promise<Favourite_Content[]>((resolve, reject) => {
            pool.query('SELECT * FROM question_user_favourites', (error, results: RowDataPacket[]) => {
            if (error) return reject(error);
    
            resolve(results as Favourite_Content[]);
            });
        });
    }



    //delete favourite relation with questionid and userid
    deleteFavouriteRelation(question_id: number, user_id: number) {
        return new Promise<void>((resolve, reject) => {
            pool.query('DELETE FROM question_user_favourites WHERE question_id=? AND user_id=?', [question_id, user_id], (error) => {
            if (error) return reject(error);
            resolve();
            });
        });
    }


}

export const favouriteService = new Service();
