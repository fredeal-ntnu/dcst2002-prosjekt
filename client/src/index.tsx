import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, CardQuestions, CardHome, Row, Column } from './widgets';
import service, { Question } from './service';
// import { TaskList, TaskDetails, TaskEdit, TaskNew } from './task-components';


class Menu extends Component {
  render() {
    return (
      <NavBar brand="askMorgan">
        <NavBar.Search/>
        <NavBar.Link to="/login">Log in</NavBar.Link>
        <NavBar.Link to="/signup">Sign up</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  questions: Question[] = [];

  render() {
    return (
      <Card title="">
        <div className="row">
          <CardHome header="Public"
            items={[{ label: "Questions", to: "/questions" }, { label: "Tags", to: "/tags" }]}/>
          <CardQuestions header="Top Questions">
          {this.questions.map((question) => (
            <Row key={question.question_id}>
              {question.title}<br></br>{question.text}
            </Row>
          ))}
        </CardQuestions>
          
        </div> 
      </Card>
    );
  }


  createQuestion() {
    service.createQuestion().then((questions) => (this.questions = questions));
  }



    mounted() {
      service.getTopFiveQuestions().then((questions) => (this.questions = questions));
    }
}


let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        {/* <Route exact path="/signin" component={login} /> */}
        {/* <Route exact path="/signup" component={signup} /> */}
        {/* <Route exact path="/signup" component={questions} /> */}
        {/* <Route exact path="/signup" component={tags} /> */}
      </div>
    </HashRouter>,
  );
