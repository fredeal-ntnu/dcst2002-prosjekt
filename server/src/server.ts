/**
 * Web server entry point used in `npm start`.
 */

import app from './app';
import express from 'express';
import path from 'path';
import { config } from 'dotenv';

// Serve client files
app.use(express.static(path.join(__dirname, '/../../client/public')));

app.get('/api/v2/current_user', (req, res) => {
  res.send(req.user);
});

const port = 3000;
app.listen(port, () => {
  console.info(`Server running on port ${port}`);
});
