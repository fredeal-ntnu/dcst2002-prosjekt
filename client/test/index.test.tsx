import React from 'react';
import { shallow } from 'enzyme';
import { Menu, Home } from 'src/index'; // Adjust the import path as needed
import { Column, Button } from '../src/widgets';
import { User } from 'src/service';
import { createHashHistory } from 'history';

const history = createHashHistory();
jest.mock('../src/service', () => {
  class Service {
    getMe() {
      return new Promise((resolve, reject) => {
        resolve([{ user_id: 1, google_id: 'test', username: 'test', email: 'test' }]);
      });
    }
  }
  return new Service();
});

describe('Site renders', () => {
  test('Site renders correct', (done) => {
    const wrapper = shallow(<Menu />);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Site renders correct', (done) => {
    const wrapper = shallow(<Home />);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
