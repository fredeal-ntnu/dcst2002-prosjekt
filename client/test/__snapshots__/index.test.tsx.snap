// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site renders Site renders correct 1`] = `
<Fragment>
  <NavBar
    brand="askMorgan"
  >
    <ButtonSuccess
      onClick={[Function]}
    >
      Login
    </ButtonSuccess>
  </NavBar>
</Fragment>
`;

exports[`Site renders Site renders correct 2`] = `
<Row>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "favourites",
          },
          {
            "label": "New Question",
            "to": "createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
  </Column>
  <Column>
    <Card
      title=""
    >
      <h3>
        Welcome!
      </h3>
      <h6>
        At 
        <b>
          askMorgan
        </b>
        , every question counts. 
        <br />
         Get answers, gain knowledge, and join our community of learners.
      </h6>
      <br />
      <h6>
        What would you like to know?
      </h6>
      <img
        height="50%"
        src="https://scontent.xx.fbcdn.net/v/t1.15752-9/363534368_210845998630767_4305768850138656455_n.jpg?_nc_cat=100&ccb=1-7&_nc_sid=510075&_nc_ohc=W06aoFJdp0EAX_ECDQp&_nc_ht=scontent.xx&oh=03_AdRfgzRf_DqEK2Ff3HU-osJWu5NMu4QuuGA5PJfyQbFQRg&oe=657C122A"
        width="50%"
      />
    </Card>
  </Column>
</Row>
`;
